src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResult.cs

```diff
@@ -0,0 +1,208 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Stores the results of analyzer execution:
+    /// 1. Local and non-local diagnostics, per-analyzer.
+    /// 2. Analyzer execution times, if requested.
+    /// </summary>
+    internal partial class AnalysisResult
+    {
+        private readonly object _gate = new object();
+        private readonly Dictionary<DiagnosticAnalyzer, TimeSpan> _analyzerExecutionTimeOpt;
+
+        private Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, List<Diagnostic>>> _localSemanticDiagnosticsOpt = null;
+        private Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, List<Diagnostic>>> _localSyntaxDiagnosticsOpt = null;
+        private Dictionary<DiagnosticAnalyzer, List<Diagnostic>> _nonLocalDiagnosticsOpt = null;
+        
+        public AnalysisResult(bool logAnalyzerExecutionTime, ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            _analyzerExecutionTimeOpt = logAnalyzerExecutionTime ? CreateAnalyzerExecutionTimeMap(analyzers) : null;
+        }
+
+        private static Dictionary<DiagnosticAnalyzer, TimeSpan> CreateAnalyzerExecutionTimeMap(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            var map = new Dictionary<DiagnosticAnalyzer, TimeSpan>(analyzers.Length);
+            foreach (var analyzer in analyzers)
+            {
+                map[analyzer] = default(TimeSpan);
+            }
+
+            return map;
+        }
+
+        public TimeSpan GetAnalyzerExecutionTime(DiagnosticAnalyzer analyzer)
+        {
+            Debug.Assert(_analyzerExecutionTimeOpt != null);
+
+            lock (_gate)
+            {
+                return _analyzerExecutionTimeOpt[analyzer];
+            }
+        }
+
+        public void Update(AnalysisScope analysisScope, AnalyzerDriver driver)
+        {
+            foreach (var analyzer in analysisScope.Analyzers)
+            {
+                // Get reported diagnostics and add them to our local diagnostic caches.
```

Comment:
In that case the comment needs rewording, because the maps are not caches and the code is dealing with non-local diagnostics as well as local ones.